pkg/resource/api.go

```diff
@@ -47,90 +47,118 @@ func NewAPIPatchingApplicator(c client.Client) *APIPatchingApplicator {
 	return &APIPatchingApplicator{client: c}
 }
 
+// WithLogger sets the logger on the APIPatchingApplicator.
+func (a *APIPatchingApplicator) WithLogger(l logging.Logger) *APIPatchingApplicator {
+	a.optionalLog = l
+	return a
+}
+
 // Apply changes to the supplied object. The object will be created if it does
 // not exist, or patched if it does. If the object does exist, it will only be
 // patched if the passed object has the same or an empty resource version.
-func (a *APIPatchingApplicator) Apply(ctx context.Context, o client.Object, ao ...ApplyOption) error {
-	m, ok := o.(metav1.Object)
-	if !ok {
-		return errors.New("cannot access object metadata")
-	}
-
-	if m.GetName() == "" && m.GetGenerateName() != "" {
-		return errors.Wrap(a.client.Create(ctx, o), "cannot create object")
+func (a *APIPatchingApplicator) Apply(ctx context.Context, obj client.Object, ao ...ApplyOption) error { //nolint:gocyclo // the logic here is crucial and deserves to stay in one method
+	if obj.GetName() == "" && obj.GetGenerateName() != "" {
+		return errors.Wrap(a.client.Create(ctx, obj), "cannot create object")
```

Comment:
in case of error, would be nice to give some details about `obj` that could not be created?

```go
return errors.Wrapf(a.client.Create(ctx, obj), "cannot create object %s of %s", obj.GetGenerateName(), obj.GetObjectKind().GroupVersionKind())
```

Reply:
Had added that context and removed it again because those error messages already contain the GR and name.